src/Raven.Server/Web/System/AdminDatabasesHandler.cs

```diff
@@ -1299,5 +1306,115 @@ public Task CompactDatabase()
             }
         }
+
+        [RavenAction("/admin/offline-migrate", "POST", AuthorizationStatus.ClusterAdmin)]
+        public async Task MigrateDatabaseOffline()
+        {
+            OfflineMigrationConfiguration configuration;
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var migrationConfiguration = await context.ReadForMemoryAsync(RequestBodyStream(), "migration-configuration");
+                configuration = JsonDeserializationServer.OfflineMigrationConfiguration(migrationConfiguration);
+            }
+                
+            var dataDir = configuration.DataDirectory;
+            if(Directory.Exists(dataDir) == false)
+                throw new DirectoryNotFoundException($"Could not find directory {dataDir}");
+
+            //var timeout = GetTimeSpanQueryString("timeout");
+            var dataExporter = configuration.DataExporterFullPath;
+            var databaseName = configuration.DatabaseName;
+            var database = await ServerStore.DatabasesLandlord.TryGetOrCreateResourceStore(databaseName, true);
+            if (database == null)
+            {
+                throw new DatabaseDoesNotExistException($"Can't import into database {databaseName} because it doesn't exist.");
+            }
+            var commandline = configuration.GenerateExporterCommandLine();
+            var processStartInfo = new ProcessStartInfo(dataExporter, commandline);
+            Stopwatch timeout = null;
+            if (configuration.Timeout.HasValue)
+            {
+                timeout = Stopwatch.StartNew();
+            }
+            processStartInfo.RedirectStandardOutput = true;
+            var processDone = new AsyncManualResetEvent();
+            var process = new Process
+            {
+                StartInfo = processStartInfo,
+                EnableRaisingEvents = true,
+            };
+            var token = new OperationCancelToken(database.DatabaseShutdown);
+            process.Exited += (sender, e) => { processDone.Set(); };
+            process.Start();
+            var progress = new IndeterminateProgress();
+            var result = new SmugglerResult();
+            var operationId = database.Operations.GetNextOperationId();
+            var timer = new Stopwatch();
+
+            await database.Operations.AddOperation(database, $"Migration of {dataDir} to {databaseName}", Documents.Operations.Operations.OperationType.DatabaseExport,
+                Onprogress =>
+                {
+                    return Task.Run(async () =>
+                    {
+                        try
+                        {
+                            timer.Start();
+                            while (processDone.IsSet == false)
+                            {
+                                progress.Progress +=$"{await process.StandardOutput.ReadLineAsync()}{Environment.NewLine}";
```

Comment:
This can wait forever, meanings that the timeout logic is not valid.
Need to handle this so you are waiting on the read task itself.